boot/modeenv.go

```diff
@@ -195,3 +204,47 @@ func splitModeenvStringList(v string) []string {
 func asModeenvStringList(v []string) string {
 	return strings.Join(v, ",")
 }
+
+// deepEquals compares two modeenvs to ensure they are textually the same. It
+// does not consider whether the modeenvs were read from disk or created purely
+// in memory. It also does not sort or otherwise mutate any sub-objects,
+// performing simple strict verification of sub-objects.
+func (m *Modeenv) deepEquals(m2 *Modeenv) bool {
+	// compare all the fields to be deep equals
+
+	// simple strings
```

Comment:
My worry with this approach is that when we add a field to the Modeenv struct we need to remember to update this helper too. Could we do something like:
```
func (m *Modeenv) deepEquals(m2 *Modeenv) bool {
	j1, err := json.Marshal(m)
	if err != nil {
		// XXX: should never happen, but handle error
	}
	j2, err := json.Marshal(m2)
	if err != nil {
		// XXX: should never happen, but handle error
	}
	return bytes.Equal(j1, j2)
}
```
instead ? It will be a bit slower but that should not be of much concern. Plus it will only compare exported parts of the struct but that is what this hellper is also doing. Or am I missing something?

Edit: I see that below json is used to check if fields added, why not using the json directly here (no strong opinion either way, just curious :)

Reply:
Yes I think you're right that just using json.Marshal here is a better idea long term to avoid bitrot in this function and then we don't have the problem that changes to the modeenv struct need to be changed here.